### PR TITLE
Error handling for User Management exposed via daml-script

### PR DESCRIPTION
**Pull request by nickchapman-da**
For #11997

The new interface is: (along with the `On` participant variants)

```
data UserId -- opaque

data User = User
  with
    id : UserId
    primaryParty : Optional Party
  deriving (Show, Eq)

exception InvalidUserId
exception UserAlreadyExists
exception UserNotFound

validateUserId : HasCallStack => Text -> Script UserId
createUser : HasCallStack => User -> [UserRight] -> Script ()
getUser : HasCallStack => UserId -> Script User
listUsers : Script [User]
grantUserRights : HasCallStack => UserId -> [UserRight] -> Script [UserRight]
revokeUserRights : HasCallStack => UserId -> [UserRight] -> Script [UserRight]
deleteUser : HasCallStack => UserId -> Script ()
listUserRights : HasCallStack => UserId -> Script [UserRight]
submitUser : HasCallStack => UserId -> Commands a -> Script a
```

### CODE REVIEW

File: daml-script/daml/Daml/Script.daml [hunk 1]
```diff
@@ -772,32 +786,78 @@ data UserRight
...
+  where
+    message (userName userId)
+
+checkUserNotFound : UserId -> Script (Optional a) -> Script a
+checkUserNotFound userId script = do
+  script >>= \case
+    None -> throw (UserNotFound userId)
+    Some x -> pure x
+
+-- | HIDE Validate some text as a user-id
```
Comment (cocreature) [inline daml-script/daml/Daml/Script.daml:828]:
```suggestion
-- | HIDE Convert the text to a user id or fail if it does not conform to the format restriction.
```

File: compiler/damlc/tests/src/DA/Test/ScriptService.hs [hunk 2]
```diff
@@ -932,70 +932,83 @@ main =
...
+                  , "  expectUserNotFound (listUserRights nonexistent)"
+                  , "  expectUserNotFound (revokeUserRights nonexistent [])"
+                  , "  expectUserNotFound (grantUserRights nonexistent [])"
                   , "  pure ()"
                   ]
                 expectScriptSuccess rs (vr "testUserManagement") $ \r ->
                     matchRegex r "Active contracts: \n"
                 expectScriptSuccess rs (vr "testUserRightManagement") $ \r ->
                     matchRegex r "Active contracts: \n"
-                expectScriptFailure rs (vr "testUserAlreadyExists") $ \r ->
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/ScriptService.hs:995]: I think we can also kill it from the proto again with this change. Let’s do that in a separate PR.

**Review by cocreature — APPROVED** [review]
Nice work thanks!

File: daml-script/daml/Daml/Script.daml [hunk 3]
```diff
@@ -506,7 +513,7 @@ instance CanAssert Script where
 data LedgerValue = LedgerValue {}
 
 fromLedgerValue : LedgerValue -> a
-fromLedgerValue = error "foobar"
+fromLedgerValue = error "foobar" --NICK: what?
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-script/daml/Daml/Script.daml]: added a more helpful comment!

File: daml-script/daml/Daml/Script.daml [hunk 4]
```diff
@@ -758,10 +765,17 @@ submitTreeMulti actAs readAs cmds =
     locations = getCallStack callStack
     continue = identity
 
+
+-- | HIDE Identifier for a user in the user management service
+data UserId = UserId
+  with
+    userName : Text
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-script/daml/Daml/Script.daml:772]: I'll do this in a following PR.

File: daml-script/daml/Daml/Script.daml [hunk 5]
```diff
@@ -772,32 +786,78 @@ data UserRight
...
+-- | HIDE Validate some text as a user-id
+validateUserId : HasCallStack => Text -> Script UserId
+validateUserId name = mkUserId name $ lift $ Free $ ValidateUserId ValidateUserIdPayload with
+  continue = pure
+  locations = getCallStack callStack
+  name
+
 -- | HIDE Create a user with the given rights
-createUser : HasCallStack => User -> [UserRight] -> Script User
+createUser : HasCallStack => User -> [UserRight] -> Script ()
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-script/daml/Daml/Script.daml:836]: After discussion, we agreed to keep the return type as`Script ()`.

File: daml-script/daml/Daml/Script.daml (hunk 1 above)
Comment (nickchapman-da) [reply daml-script/daml/Daml/Script.daml]: done

File: compiler/damlc/tests/src/DA/Test/ScriptService.hs (hunk 2 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/ScriptService.hs:995]: Ok. Will do in a separate PR.
